subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,698 @@
+) error {
+	bsl.mu.Lock()
+	defer bsl.mu.Unlock()
+
+	if err := bsl.prepare(); err != nil {
+		return err
+	}
+	bsl.run(ctx, requiresCloudCredentials, requiresWiFiCredentials, inputChan)
+
+	return nil
+}
+
+// stop stops advertising a bluetooth service which (when enabled) accepts WiFi and Viam cloud config credentials.
+func (bsl *bluetoothServiceLinux) stop() error {
+	bsl.mu.Lock()
+	defer bsl.mu.Unlock()
+
+	if !bsl.advActive {
+		return errors.New("invalid request, advertising already inactive")
+	}
+	if bsl.adv == nil {
+		return errors.New("advertisement is nil")
+	}
+	if err := bsl.adv.Stop(); err != nil {
+		return fmt.Errorf("failed to stop advertising: %w", err)
+	}
+	bsl.advActive = false
+	bsl.logger.Debug("Stopped advertising bluetooth service.")
+
+	bsl.cancelFunc()
+	bsl.workers.Wait()
+	return nil
+}
+
+func (bsl *bluetoothServiceLinux) healthy() bool {
+	bsl.mu.Lock()
+	defer bsl.mu.Unlock()
+
+	return !bsl.advActive || // If we are not advertising, we're healthy (nothing to check).
+
+		(bsl.updateAvailableWiFiNetworksHealth.IsHealthy() && // Otherwise, return combined health of goroutines.
+			bsl.listenForCredentialsHealth.IsHealthy() &&
+			bsl.listenForPairRequestHealth.IsHealthy())
+}
+
+// ---------------------------------------------------------------------------------------
+// ---------------------------------- INTERNAL METHODS -----------------------------------
+// ---------------------------------------------------------------------------------------
+
+// prepare initializes bluetooth services and defines in-memory state for storing user input.
+func (bsl *bluetoothServiceLinux) prepare() error {
+	// Can only call this function if the bsl.mu is already locked!
+
+	if bsl.advActive {
+		return errors.New("invalid request, advertising already active")
+	}
+
+	// Define each of the following characteristics in memory (and return their associated configs)
+	ssidCharacteristicConfig := initializeWriteOnlyBluetoothCharacteristic(bsl, ssidKey, 0x2222)
+	pskCharacteristicConfig := initializeWriteOnlyBluetoothCharacteristic(bsl, pskKey, 0x3333)
+	robotPartKeyIDCharacteristicConfig := initializeWriteOnlyBluetoothCharacteristic(bsl, robotPartKeyIDKey, 0x4444)
+	robotPartKeyCharacteristicConfig := initializeWriteOnlyBluetoothCharacteristic(bsl, robotPartKeyKey, 0x5555)
+	appAddressCharacteristicConfig := initializeWriteOnlyBluetoothCharacteristic(bsl, appAddressKey, 0x6666)
+	availableWiFiNetworksCharacteristicConfig := initializeReadOnlyBluetoothCharacteristic(bsl, availableWiFiNetworksKey, 0x7777)
+
+	// Create a bluetooth service comprised of the above configs.
+	serviceUUID, defaultAdvertisement, err := initializeBluetoothService(
+		bsl.deviceName,
+		[]bluetooth.CharacteristicConfig{
+			ssidCharacteristicConfig,
+			pskCharacteristicConfig,
+			robotPartKeyIDCharacteristicConfig,
+			robotPartKeyCharacteristicConfig,
+			appAddressCharacteristicConfig,
+			availableWiFiNetworksCharacteristicConfig,
+		},
```

Comment:
I think this is missing some key fields. Ideally BT provisioning should provide the same info (and therefore UI flow) as wifi. Look at line 40 of grpc.go for the list there that's sent to the mobile device. Manufacturer and Model are probably okay to skip, but FragmentID is critical for provisioning. Similarly "HasSmartMachineCredentials" and "IsOnline" values inform the mobile UI what fields to show (or not) when asking the user for input. There's also a list of errors provided, so when, say, a bad wifi password is provided (or something else goes wrong) it can be reported to the user.